enhancements/microshift/microshift-updateability-ostree.md

```diff
@@ -0,0 +1,616 @@
+and interactions with GreenBoot and operating system.
+
+## Motivation
+
+MicroShift is working towards a general availability (GA) release.
+As GA product, it is expected that it can be updated to
+provide security patches, functional updates, and bug fixes
+without needing to redeploy.
+
+MicroShift is intended to be a part of Red Hat Device Edge
+which is based on RHEL For Edge which features ostree and as such
+provides rollbacks to go back to previous ostree deployment.
+Even though, OpenShift does not support downgrade or rollback,
+MicroShift must support it in some form. Explicit downgrade 
+will not be supported, and rollback will be supported only when
+newer ostree commits is unhealthy and backup consistent with previously
+ran MicroShift is present.
+
+To allow for such operations, we need to define how we'll
+achieve that goal. We can define several areas we need to focus on:
+backing up and restoring MicroShift's data, handling version changes
+and its consequences such as migrating underlying data between schema
+versions, defining a mechanism for allowing or blocking upgrades between
+certain version of MicroShift.
+
+### User Stories
+
+* As a MicroShift administrator, I want to safely update MicroShift
+  so that I can get bug fixes, new features, and security patches.
+* As a MicroShift administrator, I want automated backups of 
+  MicroShift data so in case of a greenboot failure system will rollback
+  and backup will be restored.
+
+### Goals
+
+Goal of the enhancement is to describe implementation roadmap for
+integrating MicroShift with ostree and greenboot in order to provide
+functionality to:
+- Safely update MicroShift version (by backing up the data and
+  restoring it in case of rollback)
+- Migrating internal data (like Kubernetes storage or etcd schema) to
+  newer version
+- Block upgrades in case of version bump being too big
+
+Design aims to implement following principles:
+- Keep it simple, optimize later
+- MicroShift does not own the OS or host
+- MicroShift and all its components are version, upgraded, and rolled back together
+- Be defensive, fail fast
+- Rely on outside intervention as a last resort
+
+### Non-Goals
+
+* Building allowed/blocked version migration graph
+* Handling readiness, and backup and rollback of 3rd party applications
+  (although end user documentation should be provided)
+* Defining updateability for non-ostree systems is left to a future enhancement
+* Protecting against data corruption
+
+## Proposal
+
+### Workflow Description
+
+#### User perspective
+
+**MicroShift administrator** is a human responsible for preparing
+ostree commits and scheduling devices to use these commits.
+
+1. MicroShift administrator prepares a new ostree commit
+1. MicroShift administrator schedules device to reboot and use new ostree commit
+1. Device boots to new commit
+1. Operating System, greenboot, and MicroShift take actions without any additional intervention
```

Comment:
MicroShift admin may also decide to explicitly roll a device back. Should we describe that flow?

Reply:
This is downgrade and we don't support it

Reply:
I'd like to challenge that assumption because `rollback` is the only tool for an admin to quickly get the device working. The failure may not be related to MicroShift at all. I think we may want to support it, or it may be very limiting to administer devices.

Reply:
Sure. So biggest thing is requiring 1 backup per commit (so at least and usually 2?)

Reply:
Yes, I know we had this discussion and the decision was to go with one backup. 
However, having reviewed the proposal, I think we may want to tie a backup with a base-deployment-ID even if we keep a single one. Admins can do many things with the ostree file system (including rebases, etc.) and we need to be sure that we attempt a restore from the "right" backup.

Reply:
Sounds reasonable. @dhellmann @DanielFroehlich WDYT?

Reply:
So this is a "downgrade" with existing backup, we still won't support real downgrade (migration storage to older)

Reply:
I wouldn't call this "downgrade" - I would call this "manual rollback", even after successfull startup of a new version. Imagine e.g. problems with the workload, but a greenboot check for that is missing, then an operator might decide to manually rollback to the previous commit. We should support "manual rollbacks".

Reply:
Unless backup compatible with older version exists, from technical perspective it's a downgrade.
Looks like we're leaning toward having a backup for each ostree commit?

Reply:
We could provide more automation around this by pluging into ostree flow with `Before=ostree-finalize-staged.service`. It could check if a backup exists for staged deployment and set action to "restore"

Reply:
Automatic detection sounds like a good idea for a future enhancement.